finetuner/tuner/dataset/samplers.py

```diff
@@ -0,0 +1,220 @@
+                'batch_size must be a multiple of num_items_per_class,'
+                f' got batch_size={batch_size} and'
+                f' num_items_per_class={num_items_per_class}'
+            )
+
+        self._num_items_per_class = num_items_per_class
+        self._batch_size = batch_size
+        self._num_classes = batch_size // num_items_per_class
+
+        if len(set(labels)) < self._num_classes:
+            raise ValueError(
+                'Number of distinct classes in labels must be larger or equal than'
+                'numbers of classes in a batch'
+            )
+
+        # Get mapping of labels (classes) and their positions
+        self._class_to_labels = defaultdict(list)
+        for idx, label in enumerate(labels):
+            self._class_to_labels[label].append(idx)
+
+        # Get class groups (ids to use in batches)
+        self._cls_group_counts = []
+        for key, val in self._class_to_labels.items():
+            self._cls_group_counts.append([key, ceil(len(val) / num_items_per_class)])
+
+        # Approximate (max) length
+        self._length = ceil(
+            sum([x[1] for x in self._cls_group_counts]) / self._num_classes
+        )
+
+    def __iter__(self) -> Iterator[Tuple[int, ...]]:
+
+        # Shuffle class groups into batches, some cutoff may occur
+        counts = deepcopy(self._cls_group_counts)
+        group_batches = []
+        while len(counts) >= self._num_classes:
+            indices_batch = sample(range(len(counts)), self._num_classes)
+
+            group_batch = [counts[ind][0] for ind in indices_batch]
+            group_batches.append(group_batch)
+
+            del_inds = []
+            for ind in indices_batch:
+                counts[ind][1] -= 1
+                if counts[ind][1] == 0:
+                    del_inds.append(ind)
+
+            for ind in sorted(del_inds, reverse=True):
+                del counts[ind]
+
+        # Shuffle all labels within class, get extra samples to fill the batch
+        class_to_labels = deepcopy(self._class_to_labels)
+        for key in list(class_to_labels.keys()):
+            key_labels = class_to_labels[key]
+
+            missing_items = (
+                self._num_items_per_class - len(key_labels) % self._num_items_per_class
+            )
+            if missing_items != self._num_items_per_class:
+                key_labels += choices(key_labels, k=missing_items)
+            shuffle(key_labels)
+
+        # Construct batches
+        batches = []
+        for group_batch in group_batches:
+            batch = []
+            for cls_group in group_batch:
+                cls_items = class_to_labels[cls_group][-self._num_items_per_class :]
+                del class_to_labels[cls_group][-self._num_items_per_class :]
+                batch += cls_items
+
+            batches.append(batch)
+
+        # Update length to actual one
+        self._length = len(batches)
+
+        return iter(batches)
```

Comment:
this iter looks very complex, can you briefly tell how it works? besides, does it means that for every time we call `next`, it will build batches again from scratch? Do you want to make this batch generation function into an independent function and implement both `__iter__` and `__next__` while in `__iter__` you yield a batch? 

This is not quite clear to me, So I left a comment

Reply:
So, here's the thing: the `__iter__` methods gets called every time you do something like this
```python
for x in my_dataset: # <- This is what `__iter__` is implemented on 
    pass
```

It only gets called on the beginning of the iteration, and the `next` gets called on what is returned by `__iter__` (to be precise: here the dataset is an **iterable** and the `iter(batches)` that it returns is an **iterator**, check [python glossary](https://docs.python.org/3/glossary.html).

Now, why did I do it like this? Why not just construct `batches` in `__init__`? because in the training loop we have

```
for epoch in epochs:
    for sample in batch_sample: # Actually dataloader, but it calls sampler internally
```
And since `__iter__` will be called every epoch, with this I achieve a new shuffling of the batches every epoch.

Reply:
This is what default pytorch samplers do as well btw

Reply:
okay I see